libraries/lib-uuid/Uuid.cpp

```diff
@@ -0,0 +1,211 @@
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+/*!********************************************************************
+
+ @class Uuid
+ @brief Platform independent class for generating and parsing UUIDs.
+
+ **********************************************************************/
+
+#include "Uuid.h"
+
+#include <algorithm>
+#include <cstring>
+
+#ifdef USE_UUID_CREATE
+#   include <rpc.h>
+#endif
+
+#ifdef USE_CFUUID
+#   include <CoreFoundation/CFUUID.h>
+#endif
+
+#ifdef USE_LIBUUID
+#   include <uuid/uuid.h>
+#endif
+
+#include "lib-string-utils/HexHelpers.h"
+
+namespace audacity
+{
+
+namespace
+{
+bool readByte (
+    Uuid::Bytes::iterator& outputIt, 
+    std::string::const_iterator& inputIt, 
+    const std::string::const_iterator& inputEnd)
+{
+    if (inputIt == inputEnd)
+        return false;
+
+    const char c1 = *inputIt++;
+
+    if (!IsHexChar (c1))
+        return false;
+
+    if (inputIt == inputEnd)
+        return false;
+
+    const char c2 = *inputIt++;
+
+    if (!IsHexChar (c2))
+        return false;
+
+    *outputIt = uint8_t (
+        (HexCharToNum (c1) << 4) | 
+        HexCharToNum (c2)
+    );
+
+    ++outputIt;
+
+    return true;
+}
+}
+
+Uuid::Uuid ()
+    : Uuid (Bytes {})
+{
+}
+
+Uuid::Uuid (const Bytes& data) noexcept
+    : mData (data)
+{
+}
+
+Uuid Uuid::Generate ()
+{
+#ifdef USE_UUID_CREATE
+    UUID winUid;
+
+    if (RPC_S_OK != ::UuidCreate (&winUid))
+        return {};
+
+    Uuid uuid;
+
+    std::memcpy (uuid.mData.data (), &winUid, sizeof (winUid));
+
+    return uuid;
+#endif
+
+#ifdef USE_CFUUID
+    CFUUIDRef newId = CFUUIDCreate (NULL);
+    CFUUIDBytes bytes = CFUUIDGetUUIDBytes (newId);
+    CFRelease (newId);
+
+    Uuid uuid;
+
+    std::memcpy (uuid.mData.data (), &bytes, sizeof (bytes));
+
+    return uuid;
+#endif
+
+#ifdef USE_LIBUUID
+    uuid_t newId;
+
+    uuid_generate (newId);
+
+    Uuid uuid;
+
+    std::memcpy (uuid.mData.data (), newId, sizeof (newId));
+
+    return uuid;
+#endif
+}
+
+Uuid Uuid::FromString (const std::string& str)
+{
+    const size_t length = str.length ();
+
+    if (length == 0)
+        return {};
+
+    const bool hasBraces = str[0] == '{';
+
+    if (hasBraces && (length != 38 || str.back() != '}'))
+        return {};
+    else if (!hasBraces && length != 36)
+        return {};
+
+    std::string::const_iterator currentSymbol = hasBraces ? str.begin () + 1 : str.begin ();
+    std::string::const_iterator inputEnd = str.end ();
+
+    Uuid uuid;
+    Bytes::iterator currentByte = uuid.mData.begin ();
+
+    for (int i = 0; i < 16; ++i)
+    {
+        if (!readByte (currentByte, currentSymbol, inputEnd))
+            return {};
+
+        if (currentSymbol != inputEnd && *currentSymbol == '-')
+            ++currentSymbol;
+    }
+
+    return uuid;
+}
+
+bool Uuid::isValid () const noexcept
+{
+    return std::any_of (mData.begin (), mData.end (), [](uint8_t c) { return c != 0; });
```

Comment:
This is not a very thorough UUID verification. The minimum would be to check if the UUID version field is within the currently defined values. (1, 2, 3, 4, 5)

Reply:
It's also not correct, as zero bytes are very much allowed in UUIDs.

Reply:
It's using `std::any_of`, not `std::all_of`.  Since it's using `any_of`, it's just making sure that one of the bytes is not zero, i.e. it's not all zeros.